orderbook/src/api/order_validation.rs

```diff
@@ -0,0 +1,734 @@
+///     user sell balances or fee sufficiency.
+///     Specifically, pre-validation verifies
+///     - if buy token is native asset, receiver is not a smart contract,
+///     - the sender is not a banned user,
+///     - the order validity is appropriate,
+///     - buy_token is not the same as sell_token,
+///     - buy and sell token destination and source are supported.
+/// 2. Post-Validation:
+///     This is the additional validation performed at the time of order placement
+///     (i.e. once all the required fields on an Order are provided). Specifically, verifying that
+///     - buy & sell amounts are non-zero,
+///     - order's owner matches the from field (if specified),
+///     - fee is sufficient,
+///     - buy & sell tokens passed "bad token" detection,
+///     - user has sufficient (transferable) funds to execute the order.
+///
+///     Furthermore, Post-order validation also calls pre-validate to ensure that
+///     other aspects of the order are not malformed.
+pub struct OrderValidator {
+    /// For Pre-Validation: performed during fee & quote phase
+    code_fetcher: Box<dyn CodeFetching>,
+    native_token: WETH9,
+    banned_users: Vec<H160>,
+    min_order_validity_period: Duration,
+    /// For Post-Validation: performed time of order placement
+    fee_validator: Arc<dyn MinFeeCalculating>,
+    bad_token_detector: Arc<dyn BadTokenDetecting>,
+    balance_fetcher: Arc<dyn BalanceFetching>,
+}
+
+#[derive(Default)]
+pub struct PreOrderData {
+    pub owner: H160,
+    pub sell_token: H160,
+    pub buy_token: H160,
+    pub receiver: H160,
+    pub valid_to: u32,
+    pub buy_token_balance: BuyTokenDestination,
+    pub sell_token_balance: SellTokenSource,
+}
+
+impl From<Order> for PreOrderData {
+    fn from(order: Order) -> Self {
+        Self {
+            owner: order.order_meta_data.owner,
+            sell_token: order.order_creation.sell_token,
+            buy_token: order.order_creation.buy_token,
+            receiver: order.actual_receiver(),
+            valid_to: order.order_creation.valid_to,
+            buy_token_balance: order.order_creation.buy_token_balance,
+            sell_token_balance: order.order_creation.sell_token_balance,
+        }
+    }
+}
+
+impl OrderValidator {
+    pub fn new(
+        code_fetcher: Box<dyn CodeFetching>,
+        native_token: WETH9,
+        banned_users: Vec<H160>,
+        min_order_validity_period: Duration,
+        fee_validator: Arc<dyn MinFeeCalculating>,
+        bad_token_detector: Arc<dyn BadTokenDetecting>,
+        balance_fetcher: Arc<dyn BalanceFetching>,
+    ) -> Self {
+        Self {
+            code_fetcher,
+            native_token,
+            banned_users,
+            min_order_validity_period,
+            fee_validator,
+            bad_token_detector,
+            balance_fetcher,
+        }
+    }
+
+    pub async fn pre_validate(&self, order: PreOrderData) -> Result<(), PreValidationError> {
```

Comment:
Still not sure about the distinction here. Also this is public but the function is never used outside of this module. Will it be used later?

Reply:
This method will only be called explicitly by the fee and quote which is inside this api module. I guess it should be private then? Although it's not being called directly from within this file.

Reply:
More specifically, if this PR lands before #1155, then I would be calling it from there. I can make it private just for now and then open it up as public if/when necessary.